In [1]:
import pandas as pd

In [3]:
users = pd.read_csv(r"C:\Users\TaleyaHusn\Desktop\User Engagement\User-Engagement\Taleya - User Engagement\users.csv")
user_designations = pd.read_csv(r"C:\Users\TaleyaHusn\Desktop\User Engagement\User-Engagement\Taleya - User Engagement\user_designations.csv")
geo_tables = pd.read_csv(r"C:\Users\TaleyaHusn\Desktop\User Engagement\User-Engagement\Taleya - User Engagement\geo_tables.csv")
ahoy_visits = pd.read_csv(r"C:\Users\TaleyaHusn\Desktop\User Engagement\User-Engagement\Taleya - User Engagement\ahoy_visits.csv")
ahoy_events = pd.read_csv(r"C:\Users\TaleyaHusn\Desktop\User Engagement\User-Engagement\Taleya - User Engagement\ahoy_events.csv")

In [4]:
users.head()

id              username                       email geography_type  \
0  2649         chatna126be37         chatna126@gmail.com          Block   
1  5456   phcsalempurkala41db   phcsalempurkala@gmail.com            Phc   
2  3174       uphcsomnath0d07       uphcsomnath@gmail.com            Phc   
3  3131    phcghatoli2018c50d    phcghatoli2018@gmail.com            Phc   
4  3120  dr.bharatgoyal851297  dr.bharatgoyal85@gmail.com            Phc   

   geography_id  role_id  access_program  sign_in_count  \
0            83        3  {"",AUM,MR,TB}             12   
1           619        3          {NULL}              0   
2          1168        3         {"",MR}              1   
3          2024        3          {NULL}              0   
4          2976        3         {"",MR}              0   

           current_sign_in_at             last_sign_in_at  ...  \
0  2023-03-20 08:17:15.593453  2023-03-14 05:32:48.425878  ...   
1                         NaN                         NaN  ...   
2  2023-03-10 09:23:40.881737  2023-03-10 09:23:40.881737  ...   
3                         NaN                         NaN  ...   
4                         NaN                         NaN  ...   

                   updated_at user_designation_id  unique_session_id     org  \
0  2023-03-20 08:17:41.845943                70.0                NaN  DOMHFW   
1  2023-12-18 08:37:23.252595                 NaN                NaN     ADM   
2  2023-03-10 10:25:44.645935                52.0                NaN     NHM   
3  2023-04-10 07:05:58.205215                 NaN                NaN     NHM   
4  2023-03-09 12:11:45.222176                38.0                NaN     NHM   

  room_id                  deleted_at otp_secret  consumed_timestep  \
0     NaN                         NaN        NaN                NaN   
1     NaN                         NaN        NaN                NaN   
2     NaN                         NaN        NaN                NaN   
3     NaN  2023-04-10 07:05:58.205202        NaN                NaN   
4     NaN                         NaN        NaN                NaN   

  otp_required_for_login otp_backup_codes  
0                    NaN              NaN  
1                    NaN              NaN  
2                    NaN              NaN  
3                    NaN              NaN  
4                    NaN              NaN  

[5 rows x 26 columns]

# Checking for null values column wise for each table

### Users

In [5]:
percent_missing = users.isnull().sum()*100/len(users)
percent_missing

id                          0.000000
username                    0.000000
email                       0.000000
geography_type              0.000000
geography_id                0.000000
role_id                     0.000000
access_program              0.000000
sign_in_count               0.000000
current_sign_in_at         41.804564
last_sign_in_at            41.804564
current_sign_in_ip         41.804564
last_sign_in_ip            41.804564
failed_attempts             0.000000
unlock_token              100.000000
locked_at                  87.702491
created_at                  0.000000
updated_at                  0.000000
user_designation_id        22.609302
unique_session_id          99.477443
org                         0.243860
room_id                    95.784706
deleted_at                 84.776171
otp_secret                 98.972304
consumed_timestep          99.930326
otp_required_for_login     99.930326
otp_backup_codes          100.000000
dtype: float64

In [19]:
users['access_program']

0                                          {"",AUM,MR,TB}
1                                                  {NULL}
2                                                 {"",MR}
3                                                  {NULL}
4                                                 {"",MR}
                              ...                        
5736                            {"",DHS,AUM,Report,MR,TB}
5737                                       {"",AUM,MR,TB}
5738    {"",PW,DHS,FP,MH,CH,IDSP,AUM,NCD,Report,MR,TB,...
5739      {"",DHS,FP,MH,CH,IDSP,AUM,NCD,Report,MR,TB,FSP}
5740      {"",DHS,FP,MH,CH,IDSP,AUM,NCD,Report,MR,TB,FSP}
Name: access_program, Length: 5741, dtype: object

#### We can see here that there is no null column but, there are sum columns explicitly marked as '{NULL}'

In [5]:
null_access_program = users.loc[users['access_program'] == '{NULL}', 'id']
print(null_access_program)


1       5456
3       3131
5       3327
8       3393
13      3128
        ... 
5650    5375
5675    5144
5679    5143
5719    5269
5721    5155
Name: id, Length: 706, dtype: int64


#### There are 706 values explicitly marked as 'NULL'

## Conclusion from Users table-
>1. otp_backup_codes, otp_required_for_login, consumed_timestep,otp_secret, deleted_at, room_id, unique_session_id, locked_at, unlocked_token columns doesnot has any value (approx)
>2. current_sign_in_id, last_sign_in_ip, current_sign_in_at, last_sign_in_at has 40% null values
>3. 'access_program' has null values explicitly marked as '{Null}'

## User Designation

In [6]:
percent_missing = user_designations.isnull().sum()*100/len(user_designations)
percent_missing

id                0.0
geography_type    0.0
name_en           0.0
created_at        0.0
updated_at        0.0
dtype: float64

### Conclusion
> There are no null values in Users Designation table

## Geo_tables

In [52]:
geo_tables.head()

state_id state_name  district_id district_name  block_id   block_name  \
0         1  Rajasthan            1         Ajmer         1  Ajmer Urban   
1         1  Rajasthan            1         Ajmer         1  Ajmer Urban   
2         1  Rajasthan            1         Ajmer         1  Ajmer Urban   
3         1  Rajasthan            1         Ajmer         1  Ajmer Urban   
4         1  Rajasthan            1         Ajmer         1  Ajmer Urban   

   phc_id                phc_name  sub_center_id         sub_center_name  \
0       2  City Dispensary Beawar              2  City Dispensary Beawar   
1       2  City Dispensary Beawar              2  City Dispensary Beawar   
2       2  City Dispensary Beawar              2  City Dispensary Beawar   
3       2  City Dispensary Beawar              2  City Dispensary Beawar   
4       2  City Dispensary Beawar              2  City Dispensary Beawar   

   village_id village_name  anganwadi_id              anganwadi_name  
0          18      Ward 11          1575          Teja chok(Ward 11)  
1          18      Ward 11          1579       Damodar park(Ward 11)  
2          18      Ward 11          1583     Regranbada bas(Ward 11)  
3          19      Ward 16          1577  Maliyan Mohalla 1(Ward 16)  
4          19      Ward 16          1569          Udan chok(Ward 16)

In [7]:
percent_missing = geo_tables.isnull().sum()*100/len(geo_tables)
percent_missing

state_id           0.0
state_name         0.0
district_id        0.0
district_name      0.0
block_id           0.0
block_name         0.0
phc_id             0.0
phc_name           0.0
sub_center_id      0.0
sub_center_name    0.0
village_id         0.0
village_name       0.0
anganwadi_id       0.0
anganwadi_name     0.0
dtype: float64

### Conclusion
> No null value in geo_table

## ahoy_events

In [8]:
ahoy_events.head()

id  visit_id  user_id           name                            properties  \
0   2         1      1.0  data_explorer  {"indicator": "surveyed_population"}   
1   3         1      1.0  data_explorer  {"indicator": "surveyed_population"}   
2   4         5    673.0  Data Explorer                               "click"   
3   5         5    673.0    MR Campaign                               "click"   
4   6         6    277.0    MR Campaign                               "click"   

                         time  
0     2022-12-30 13:07:55.421  
1     2022-12-30 13:10:06.604  
2  2022-12-30 13:10:43.520913  
3  2022-12-30 13:10:57.166318  
4  2022-12-30 13:11:11.448949

In [10]:
percent_missing = ahoy_events.isnull().sum()*100/len(ahoy_events)
percent_missing

id            0.000000
visit_id      0.000000
user_id       0.258063
name          0.000000
properties    0.000000
time          0.000000
dtype: float64

## Conclusion
> 'user_id' has null values.

## ahoy_visits

In [11]:
ahoy_visits.head()

id                           visit_token  \
0      1  208e6438-2283-4de2-804c-d19f7e37ddf9   
1  35651  65f7123e-c691-40cc-a034-e7bfd62ca199   
2  35630  37ef33fc-4861-4cae-80b5-2726262f90c2   
3  35645  bb6c9cc2-3363-460d-aea3-8c657aa031bb   
4      5  78520419-4003-404b-b0f1-620995538418   

                          visitor_token  user_id              ip  \
0  778c5576-7e28-48df-8ba2-e9542be572ea      1.0   103.59.75.228   
1  4af745c0-16a1-4f21-ba21-47d15cd447aa      NaN  157.38.132.154   
2  82ec62fa-e3e9-4dd3-8dfa-92b43c85bd36    302.0  157.38.133.194   
3  8a64ff13-56a8-43e3-8694-4935fd2c7c77      NaN  223.233.121.44   
4  ac3e4aba-fe3d-4581-9e69-22e88a435815    673.0   157.38.42.107   

                                          user_agent  \
0  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; ...   
1  Mozilla/5.0 (Linux; Android 11; RMX2027) Apple...   
2  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
3  Mozilla/5.0 (Linux; Android 13; V2153) AppleWe...   
4  Mozilla/5.0 (Linux; U; Android 9; en-in; Redmi...   

                                        referrer         referring_domain  \
0  https://nirogi.rajasthan.gov.in/users/sign_in  nirogi.rajasthan.gov.in   
1                        https://www.google.com/           www.google.com   
2                        https://www.google.com/           www.google.com   
3                        https://www.google.com/           www.google.com   
4                                            NaN                      NaN   

                                        landing_page        browser  ...  \
0                   https://nirogi.rajasthan.gov.in/        Firefox  ...   
1  https://www.nirogi.rajasthan.gov.in/users/sign_in  Chrome Mobile  ...   
2  https://www.nirogi.rajasthan.gov.in/users/sign_in         Chrome  ...   
3  https://www.nirogi.rajasthan.gov.in/users/sign_in  Chrome Mobile  ...   
4               https://www.nirogi.rajasthan.gov.in/   MIUI Browser  ...   

  longitude utm_source utm_medium utm_term utm_content  utm_campaign  \
0   75.7105        NaN        NaN      NaN         NaN           NaN   
1   77.0060        NaN        NaN      NaN         NaN           NaN   
2   77.0060        NaN        NaN      NaN         NaN           NaN   
3   75.7105        NaN        NaN      NaN         NaN           NaN   
4   75.7105        NaN        NaN      NaN         NaN           NaN   

   app_version  os_version  platform                  started_at  
0          NaN         NaN       NaN  2022-12-30 13:07:48.737643  
1          NaN         NaN       NaN  2023-02-02 16:26:51.521452  
2          NaN         NaN       NaN  2023-02-02 16:13:00.554915  
3          NaN         NaN       NaN  2023-02-02 16:21:25.513858  
4          NaN         NaN       NaN  2022-12-30 13:10:43.457746  

[5 rows x 26 columns]

In [12]:
percent_missing = ahoy_visits.isnull().sum()*100/len(ahoy_visits)
percent_missing

id                    0.000000
visit_token           0.000660
visitor_token         0.000000
user_id              32.443307
ip                    0.000000
user_agent            0.000000
referrer             49.733520
referring_domain     49.749350
landing_page          0.166880
browser               0.021767
os                    0.000000
device_type           0.077833
country               1.062623
region               29.300292
city                 29.624817
latitude              1.062623
longitude             1.062623
utm_source          100.000000
utm_medium          100.000000
utm_term            100.000000
utm_content         100.000000
utm_campaign        100.000000
app_version         100.000000
os_version          100.000000
platform             93.065578
started_at            0.000000
dtype: float64

### Conclusion
> utm_medium, utm_term, utm_content, utm_campaign, app_version, os_version, has all null values.
> region, city,user_id has 30 percent null values
> referrer, referring domain has 50% null values.